src/adapter/src/coord.rs

```diff
@@ -1804,6 +1900,24 @@ impl<S: Append + 'static> Coordinator<S> {
         .expect("updating compute instance status cannot fail");
     }
 
+    #[tracing::instrument(level = "debug", skip_all)]
+    async fn message_update_read_holds(&mut self) {
+        let global_timelines = std::mem::take(&mut self.global_timelines);
```

Comment:
Could this be `for (..) in self.global_timelines.iter_mut()` instead of re allocating a hashmap each time?

Reply:
I cannot get the borrow checker to allow this. The issue being that `self.global_timelines.iter_mut()` requires a mutable reference to `self` and `self.update_read_hold` also requires a mutable reference to `self`. The specific error is below:
```
error[E0499]: cannot borrow `*self` as mutable more than once at a time
    --> src/adapter/src/coord.rs:1906:13
     |
1904 |         for (timeline, TimelineState { oracle, read_holds }) in self.global_timelines.iter_mut() {
     |                                                                 --------------------------------
     |                                                                 |
     |                                                                 first mutable borrow occurs here
     |                                                                 first borrow later used here
1905 |             let new_time = oracle.read_ts();
1906 |             self.update_read_hold(read_holds, new_time).await;
     |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ second mutable borrow occurs here

For more information about this error, try `rustc --explain E0499`.
```

If you have any ideas let me know.

Reply:
Oh right, I hit this same problem. It's fine?

Reply:
I hope so. The Coordinator has so much state that I keep running into this issue. I often want to mutably iterate through some field and for each element call a method that requires a mutable reference to `self`. My only hope is that the compiler ends up optimizing these away,